File: Codeforces/div859/D.py

def inp():
    return list(map(int, input().split()))

for _ in range(int(input())):
    n, q = inp()
    a = inp()
    pref = [0] * n
    suff = [0] * n

    pref[0] = a[0]
    for i in range(1, n):
        pref[i] = pref[i - 1] + a[i]

    suff[-1] = a[-1]
    for i in range(n - 2, -1, -1):
        suff[i] = suff[i + 1] + a[i]

    for i in range(q):
        l, r, k = inp()
        l -= 1
        r -= 1
        if l == 0:
            L = 0
        else:
            L = pref[l - 1]
        if r == n - 1:
            R = 0
        else:
            R = suff[r + 1]

        ans = L + R + k * (r - l + 1)
        if ans % 2 == 0:
            print('NO')
        else:
            print('YES')

File: Codeforces/div859/D.cpp

#include <bits/stdc++.h>
using namespace::std;
#define int long long
signed main() {
    int t;
    cin >> t;
    while (t--) {
        int n, q;
        cin >> n >> q;

        vector<int> a(n);
        for (int i = 0; i < n; ++i) {
            cin >> a.at(i);
        }
        vector<int> pref(n);
        vector<int> suff(n);

        pref.at(0) = a.at(0);
        for (int i = 1; i < n; ++i) {
            pref.at(i) = pref.at(i - 1) + a.at(i);
        }

        suff.at(n - 1) = a.at(n - 1);
        for (int i = n - 2; i >= 0; --i) {
            suff.at(i) = suff.at(i + 1) + a.at(i);
        }

        for (int i = 0; i < q; ++i) {
            int l, r, k;
            cin >> l >> r >> k;
            --l;
            --r;
            int L, R;
            if (l == 0) {
                L = 0;
            } else {
                L = pref.at(l - 1);
            }
            if (r == n - 1) {
                R = 0;
            } else {
                R = suff.at(r + 1);
            }
            int ans = L + R + k * (r - l + 1);
            if (ans % 2 == 0) {
                cout << "NO\n";
            } else {
                cout << "YES\n";
            }
        }
    }
}
